tensorflow/python/ops/custom_gradient.py

```diff
@@ -165,6 +165,90 @@ def second_order_and_transpose(ddy):
   Additional arguments to the inner `@tf.custom_gradient`-decorated function
   control the expected return values of the innermost function.
 
+  Above examples illustrates how you can create custom scalar gradient
+  function. Now, let us understand using linear polynomial on how to create
+  gradient which also computes gradient of a variable which can be `weights`
+  of model.
+
+  ```python
+  >>> trainable = []
+  ... # List to store the trainable variables (weights, biases etc.)
+  ...
+  ... @tf.custom_gradient
+  ... def linear_poly(x):
+  ...   # Weights will be trainable parameter and will represent coefficients
+  ...   # of poly and initialized as ones.
+  ...   weights = tf.Variable(tf.ones([2]))
+  ...   trainable.append(weights)
+  ...
+  ...   # Creating polynomial
+  ...   poly = weights[1] * x + weights[0]
+  ...
+  ...   def grad_fn(dpoly, variables=None):
+  ...     # dy/dx = weights[1] and we need to left multiply dpoly
+  ...     grad_xs = dpoly * weights[1]  # Scalar gradient
+  ...
+  ...     input_shape = len(x.shape)  # x contains how many dimensions
+  ...     grad_vars = []  # To store gradients of passed variables
+  ...     if variables is not None:
+  ...       for v in variables:
+  ...         if v is weights:
+  ...           # Manually computing dy/dweights
+  ...           dy_dw = tf.Variable([x ** 1, x ** 0])
+  ...           dy_dw_concat = tf.concat(dy_dw, axis=0)
+  ...           if input_shape > 0:
+  ...             # Note here that reduce operation is performed
+  ...             # in axis so that it computes gradient w.r.t. entire batch
+  ...             # examples and it doesn't reduce the entire weights Variable
+  ...             # that is why we are only reducing on axis 1.
+  ...             grad_vars.append(
+  ...                 tf.reduce_sum(dy_dw_concat, axis=1)
+  ...             )
+  ...           else:
+  ...             # No need to reduce if x is scalar
+  ...             grad_vars.append(dy_dw_concat)
+  ...         else:  # Create elif if you have more variables
+  ...           grad_vars.append(None)
+  ...     return grad_xs, grad_vars
+  ...   return poly, grad_fn
+  >>> x = tf.Variable([1., 2., 3.])
+  >>> with tf.GradientTape(persistent=True) as tape:
+  ...   poly = linear_poly(x)
+  >>> poly  # poly = x + 1
+  <tf.Tensor: shape=(3,),
+    dtype=float32,
+    numpy=array([2., 3., 4.], dtype=float32)>
+  >>> tape.gradient(poly, x)  # conventional scalar gradient dy/dx
+  <tf.Tensor: shape=(3,),
+    dtype=float32,
+    numpy=array([1., 1., 1.], dtype=float32)>
+  >>> tape.gradient(poly, trainable[0])  # gradient w.r.t weights dy/dweights
+  <tf.Tensor: shape=(2,),
+    dtype=float32,
+    numpy=array([6., 3.], dtype=float32)>
+  ```
+
+  Above example illustrates two things:
+  - Usage of trainable variable `weights` and computing it's gradient.
+  - Computing gradient in case of batches.
+
+  In example, You can see the inner `grad_fn` that accepts extra `variables`
```

Comment:
grammar/style nits: "In the example, the inner `grad_fn` accepts an extra `variables` input parameter and also returns an extra `grad_vars` output." (I'd leave off the "scalar gradient functions" bit, since it's stateless vs. stateful more than scalar vs. non-scalar)

Reply:
done